### PR TITLE
Add boolean attributes

### PR DESCRIPTION
**Pull request by jmphilli**

### CODE REVIEW

**Comment by jmphilli** [comment]
@johnliu @danielhochman @garrettheel

File: pynamodb/tests/test_model.py [hunk 1]
```diff
@@ -2413,3 +2423,104 @@ def test_loads(self):
             ]
         }
         self.assert_dict_lists_equal(req.call_args[0][1]['RequestItems']['UserModel'], args['UserModel'])
+
+    def test_old_style_boolean_serializes_as_bool(self):
+
```
Comment (garrettheel) [inline pynamodb/tests/test_model.py]: ultra nit: delete newline

File: pynamodb/tests/test_model.py [hunk 2]
```diff
@@ -2413,3 +2423,104 @@ def test_loads(self):
...
+        with patch(PATCH_METHOD) as req:
+            req.return_value = BOOLEAN_CONVERSION_MODEL_TABLE_DATA
```
Comment (garrettheel) [inline pynamodb/tests/test_model.py]: This isn't testing the old style, right? `BOOLEAN_CONVERSION_MODEL_TABLE_DATA` has `is_human` defined as a `BOOL`

File: pynamodb/models.py [hunk 3]
```diff
@@ -447,7 +448,12 @@ def from_raw_data(cls, data):
         for name, value in mutable_data.items():
             attr = cls._get_attributes().get(name, None)
             if attr:
-                kwargs[name] = attr.deserialize(value.get(ATTR_TYPE_MAP[attr.attr_type]))
+                serialized_dynamo_type = ATTR_TYPE_MAP[attr.attr_type]
+                value_to_deserialize = value.get(serialized_dynamo_type)
+                # this only needs to be here until models have all BOOL types serialized
```
Comment (garrettheel) [inline pynamodb/models.py]: You should assume that's pretty much until the end of time and just handle it gracefully with a comment about why we need to do this (legacy compat)

File: pynamodb/attributes.py [hunk 4]
```diff
@@ -186,30 +190,48 @@ def deserialize(self, value):
...
-        """
-        return bool(json.loads(value))
+        return bool(value)
+
+    def get_value(self, value):
+        # we need this for legacy compatibility.
+        # previously, BOOL was serialized as N
+        value_to_deserialize = super(BooleanAttribute, self).get_value(value)
+        if value_to_deserialize is None:
+            value_to_deserialize = int(value.get(NUMBER_SHORT))
```
Comment (danielhochman) [inline pynamodb/attributes.py]: safety first

**Comment by danielhochman** [comment]
+1 pending last two comments

File: pynamodb/attributes.py [hunk 5]
```diff
@@ -186,30 +190,48 @@ def deserialize(self, value):
...
+
+    def deserialize(self, value):
+        return bool(json.loads(value))
+
+
 class BooleanAttribute(Attribute):
     """
     A class for boolean attributes
 
     This attribute type uses a number attribute to save space
```
Comment (danielhochman) [inline pynamodb/attributes.py]: update docstring

File: pynamodb/attributes.py [hunk 6]
```diff
@@ -186,30 +190,48 @@ def deserialize(self, value):
...
     """
-    attr_type = NUMBER
+    attr_type = BOOLEAN
 
     def serialize(self, value):
-        """
-        Encodes True as 1, False as 0
-        """
         if value is None:
```
Comment (danielhochman) [inline pynamodb/attributes.py]: can this even happen?

**Pull request merged**